### PR TITLE
Update Prometheus vendoring, adapt Cortex

### PR DESCRIPTION
**Pull request by juliusv**
This is preparatory work for https://github.com/weaveworks/cortex/issues/622

The most noticeable change is around the lifecycle manager of the Alertmanager
notifier and its associated service discovery manager. This has now been moved
from the notifier package into main.go in Prometheus and thus we also need to
duplicate that lifecycle management ourselves.

### CODE REVIEW

**Comment by juliusv** [comment]
(and some test updates still needed)

**Comment by juliusv** [comment]
Ok, tests updated and additionally verified that it still sends notifications in a real setup.

@jml Should be ready now.

File: pkg/ruler/ruler.go [hunk 1]
```diff
@@ -97,7 +106,67 @@ type Ruler struct {
 
 	// Per-user notifiers with separate queues.
 	notifiersMtx sync.Mutex
-	notifiers    map[string]*notifier.Notifier
+	notifiers    map[string]*rulerNotifier
+}
+
+type rulerNotifier struct {
```
Comment (jml) [inline pkg/ruler/ruler.go]: Can you please add a comment for this explaining what it is and why we want it?

File: pkg/ruler/ruler.go [hunk 2]
```diff
@@ -191,26 +260,59 @@ func (r *Ruler) newGroup(ctx context.Context, rs []rules.Rule) (*rules.Group, er
...
-		Notifier:    notifier,
+		NotifyFunc:  sendAlerts(notifier, r.alertURL.String()),
 		Logger:      gklog.NewNopLogger(),
+		Registerer:  prometheus.DefaultRegisterer,
 	}
 	delay := 0 * time.Second // Unused, so 0 value is fine.
 	return rules.NewGroup("default", "none", delay, rs, opts), nil
 }
 
+// sendAlerts implements a the rules.NotifyFunc for a Notifier.
```
Comment (jml) [inline pkg/ruler/ruler.go]: Nit: s/a the/a/

File: pkg/ruler/ruler.go (hunk 1 above)
Comment (juliusv) [reply pkg/ruler/ruler.go]: Done.

File: pkg/ruler/ruler.go (hunk 2 above)
Comment (juliusv) [reply pkg/ruler/ruler.go]: Done.

**Pull request merged**